pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -99,20 +115,55 @@ func (p *Provider) Init() error {
 
 // Provide allows the consul catalog provider to provide configurations to traefik using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- dynamic.Message, pool *safe.Pool) error {
+	var err error
+	p.client, err = createClient(p.Endpoint)
+	if err != nil {
+		return fmt.Errorf("unable to create consul client: %w", err)
+	}
+
+	if p.ConnectAware {
+		// When using connect the certificates should be fetched using the local agent
```

Comment:
So this is a tricky one. Treafik's consul catalog connects to a consul server to fetch the service data. Connect should connect to a local agent instead. When this is launched in let's say nomad as orchestrator then nomad will provide a token & endpoint to use via env variables. This makes a lot of sense and I would rather not add configuration params for that in the catalog config. WDYT?

Reply:
No clue atm, let me Consul(t) with my coworkers.

Reply:
Are you saying that: when Traefik is orchestrated/launched by whatever thing handles a consul "cluster" (nomad?), it can be launched with specific environment variables, which will make L128 automatically work (i.e. it will make a client that is supposed to connect with an agent that we are supposed to call for all API calls related to consul)?

If yes, are you saying that is the most common case? And the case that interests you and other people involved in this PR?

If yes, I think you will find that the situation is similar to what happens in github.com/traefik/traefik/pkg/provider/kubernetes/ingress/kubernetes.go , L73-84

Assuming we understood correctly and all of the above is true , I suppose we could do either like in kubernetes and try to support all scenarii, or we could support only the first case for now. And we could add the rest later on if needed.

WDYT?

Reply:
>  it can be launched with specific environment variables, which will make L128 automatically work (i.e. it will make a client that is supposed to connect with an agent that we are supposed to call for all API calls related to consul)?

Yes, this is default behavior of any cosul client though (if the go consul sdk is used). https://github.com/hashicorp/consul/blob/b76dfa144167b5c3fa05a7f00237927ea3333a85/api/api.go#L388-L470 This config is used to supply any missing values: https://github.com/hashicorp/consul/blob/b76dfa144167b5c3fa05a7f00237927ea3333a85/api/api.go#L595 -- I just realized though that traefik overrides this default behavior: https://github.com/traefik/traefik/blob/master/pkg/provider/consulcatalog/consul_catalog.go#L68-L70 -- if I simply drop this, the defaults will be exactly the same, but can also be configured via env variables.

> If yes, are you saying that is the most common case? And the case that interests you and other people involved in this PR?

It would certainly be a usecase to support; I cannot imagine many people manually configuring a service mesh without orchestrator support.

> If yes, I think you will find that the situation is similar to what happens in github.com/traefik/traefik/pkg/provider/kubernetes/ingress/kubernetes.go , L73-84

Yes and the consul already implicitly does that -- I will rewrite and see what comes out of it.

Reply:
I force-pushed a change, because I had a condition wrong; though I am not sure if that could be the cause.

Reply:
So, with the PR as it is now, does Traefik with consul connect work "in production" for you?

Reply:
It should (tm), I'll have to create a docker container with the binary and try in my cluster; but local tests look good.

Reply:
ok, thanks. let us know please.

Reply:
So, initial tests look good; Ie I can submit the following job to nomad and it works:
```
job "test" {
  datacenters = ["dc1"]
  namespace = "default"

  group "c1" {
    network {
      port "p1" {}
    }

    service {
      name = "uuid-api"
      port = "p1"
      tags = [
          "traefik.enable=true",
          "traefik.connect=true",
          "traefik.http.routers.backend1.rule=PathPrefix(`/`)"
      ]
      connect {
        native = true
      }
    }

    task "c1" {
      driver = "docker"

      config {
        image        = "hashicorpnomad/uuid-api:v5"
        network_mode = "host"
      }

      env {
        BIND = "0.0.0.0"
        PORT = "${NOMAD_PORT_p1}"
        UPSTREAM = "backend1"
      }
    }
  }

  group "traefik" {
    network {
      port "p1" {}
    }

    service {
      name = "traefik"
      port = "p1"
      connect {
        native = true
      }
    }

    task "c1" {
      driver = "docker"

      config {
        image        = "traefik/traefik:connect"
        network_mode = "host"
      }

      env {
        TRAEFIK_ENTRYPOINTS_HTTP_ADDRESS="0.0.0.0:${NOMAD_PORT_p1}"
        TRAEFIK_PROVIDERS_CONSULCATALOG="true"
        TRAEFIK_PROVIDERS_CONSULCATALOG_CONNECTAWARE="true"
        TRAEFIK_PROVIDERS_CONSULCATALOG_SERVICENAME="traefik"
        TRAEFIK_PROVIDERS_CONSULCATALOG_EXPOSEDBYDEFAULT="false"
        TRAEFIK_LOG_LEVEL="debug"
      }
    }
  }

}
```

On thing I did notice though during testing is that traefik always thinks that my configuration is "new", it I never get:
```
time="2021-02-26T16:29:15Z" level=info msg="Skipping same configuration" providerName=consulcatalog
```
This is because the configuration watcher uses `reflect.DeepEqual` which obviously fails for my dynamically generated certification verification function because it changes every time. Before I try to fix this I was wondering if you could provide me with some hints how that is best solved.

Reply:
hmm, good point. No idea off the top of my head, but I'm going to think about it. Maybe if we used a function pointer instead?

Reply:
ouch, I forgot about that: "Func values are deeply equal if both are nil; otherwise they are not deeply equal."

Reply:
So yeah, it seems like maybe we could cheat by using pointers, e.g. this seems to do the trick:

```
package main

import (
	"reflect"

	"github.com/davecgh/go-spew/spew"
)

type config struct {
	myfun *func() string
}

func fixedFunc() string {
	return "hello"
}

func main() {
	var prev, cur config
	thefun := fixedFunc
	cur = config{
		myfun: &thefun,
	}
	prev = cur
	spew.Dump(prev)
	spew.Dump(cur)
	println(reflect.DeepEqual(prev, cur))
}
```

Reply:
More realistic version:

```
package main

import (
	"reflect"
	"time"

	"github.com/davecgh/go-spew/spew"
)

type config struct {
	myfun *func() string
}

func fixedFunc() string {
	return "hello"
}

var thefun func() string

func init() {
	thefun = fixedFunc
}

func main() {
	var prev, cur config
	cur = config{
		myfun: &thefun,
	}
	prev = cur
	for {
		cur = config{
			myfun: &thefun,
		}
		spew.Dump(prev)
		spew.Dump(cur)
		println(reflect.DeepEqual(prev, cur))
		time.Sleep(5 * time.Second)
	}
}
```

Reply:
I just tried that, but that would require a "fixedFunc", no? The function that I have cannot be fixed unless we stash it somewhere else :/

This was my attempt at fixing:
```
diff --git a/pkg/config/dynamic/http_config.go b/pkg/config/dynamic/http_config.go
index c5c2a8fc..d5e267b8 100644
--- a/pkg/config/dynamic/http_config.go
+++ b/pkg/config/dynamic/http_config.go
@@ -214,7 +214,7 @@ type ServersTransport struct {
 	// order to implement the Consul Connect protocol between Traefik and the Consul
 	// services. Therefore it is not serializable to prevent it from being configurable
 	// by other parties.
-	VerifyPeerCertificate func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error `json:"-" toml:"-" yaml:"-" hash:"-" label:"-" export:"false"`
+	VerifyPeerCertificate *func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error `json:"-" toml:"-" yaml:"-" hash:"-" label:"-" export:"false"`
 }
 
 // +k8s:deepcopy-gen=true
diff --git a/pkg/provider/consulcatalog/config.go b/pkg/provider/consulcatalog/config.go
index aeed583e..60e96ebd 100644
--- a/pkg/provider/consulcatalog/config.go
+++ b/pkg/provider/consulcatalog/config.go
@@ -150,30 +150,33 @@ func (c *connectCert) equals(other *connectCert) bool {
 }
 
 func (c *connectCert) serverTransport(item itemData) *dynamic.ServersTransport {
+	verifier := func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// We should use RootCAs here, but consul expect that in ClientCAs (don't ask)
+		// https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L279-L282
+		// called via https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L258
+		cfg.ClientCAs = cfg.RootCAs
+		cert, err := verifyChain(cfg, rawCerts, true)
+		if err != nil {
+			return err
+		}
+		certs := []*x509.Certificate{0: cert}
+		uri := &connect.SpiffeIDService{
+			Namespace:  item.Namespace,
+			Datacenter: item.Datacenter,
+			Service:    item.Name,
+		}
+		return verifyServerCertMatchesURI(certs, uri)
+	}
 	return &dynamic.ServersTransport{
+		// This ensure that the config changes, whenever the verifier func changes as well
+		ServerName: fmt.Sprintf("%s-%s-%s", item.Namespace, item.Datacenter, item.Name),
 		// InsecureSkipVerify is needed because Go wants to verify a hostname otherwise
 		InsecureSkipVerify: true,
 		RootCAs:            c.getRoot(),
 		Certificates: tls.Certificates{
 			c.getLeaf(),
 		},
-		VerifyPeerCertificate: func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-			// We should use RootCAs here, but consul expect that in ClientCAs (don't ask)
-			// https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L279-L282
-			// called via https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L258
-			cfg.ClientCAs = cfg.RootCAs
-			cert, err := verifyChain(cfg, rawCerts, true)
-			if err != nil {
-				return err
-			}
-			certs := []*x509.Certificate{0: cert}
-			uri := &connect.SpiffeIDService{
-				Namespace:  item.Namespace,
-				Datacenter: item.Datacenter,
-				Service:    item.Name,
-			}
-			return verifyServerCertMatchesURI(certs, uri)
-		},
+		VerifyPeerCertificate: &verifier,
 	}
 }
 
diff --git a/pkg/server/service/roundtripper.go b/pkg/server/service/roundtripper.go
index fe052825..9f89e4a8 100644
--- a/pkg/server/service/roundtripper.go
+++ b/pkg/server/service/roundtripper.go
@@ -148,7 +148,7 @@ func createRoundTripper(cfg *dynamic.ServersTransport) (http.RoundTripper, error
 		}
 		if cfg.VerifyPeerCertificate != nil {
 			transport.TLSClientConfig.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-				return cfg.VerifyPeerCertificate(transport.TLSClientConfig, rawCerts, verifiedChains)
+				return (*cfg.VerifyPeerCertificate)(transport.TLSClientConfig, rawCerts, verifiedChains)
 			}
 		}
 	}
```

Reply:
Well, it does not have to be fixed (whatever that means), but as we cannot directly take the address of a function definition (we need an actual variable with a value of type function, of which we take the address), that's why I was doing this dance with a func definition and a global var to hold the value.
But yes, we want the value to stay the same from one Provide call to the other, or at least as long as nothing else changes in the configuration. So maybe we could "cache" the value of the variable that holds the function as a non-exported field of the Provider?

A coworker also suggested a solution based on interfaces, but I haven't studied it yet. I'll let you know.

Reply:
And since the function value is dependent on item.Namespace, item.Datacenter, and item.Name (which can probably be uniquely represented as a SpiffeIDService.String() right?), I think we have two solutions here:

1) Either we keep only only one (key, value) couple (SpiffeIDService.String(), func value), and whenever serverTransport is called with a different item we forget about it, the func value gets erased, and a config refresh will be triggered

2) or we keep a map of the above (key, value)s so that we never end up refreshing for an item that we have already previously seen. But I don't know whether that's reasonable yet.

Reply:
> which can probably be uniquely represented as a SpiffeIDService.String() right?

Yes, the function only has a dependency on `SpiffeIDService` because it gets the root certs from the TLS config itself, so if those change the whole config would change as well.

> Either we keep only only one (key, value) couple (SpiffeIDService.String(), func value)

But the `SpiffeIDService.String()` is for the target service. So for X services discovered, we'd have X different `SpiffeIDService` -- I do not think that will work

> or we keep a map of the above (key, value)s so that we never end up refreshing for an item that we have already previously seen. But I don't know whether that's reasonable yet.

That seems the way to go, I am just a little bit worried what to do about items that vanish from the consul server; when will that cache get cleaned up…

Reply:
yeah, hold on, we may have another solution. Let my try it out a bit on my side so I don't make you waste your time for nothing.

Reply:
here is the alternative, WDYT?

```
diff --git a/pkg/config/dynamic/http_config.go b/pkg/config/dynamic/http_config.go
index c5c2a8fcd..d9527050b 100644
--- a/pkg/config/dynamic/http_config.go
+++ b/pkg/config/dynamic/http_config.go
@@ -200,6 +200,10 @@ func (h *HealthCheck) SetDefaults() {
        h.FollowRedirects = &fr
 }

+type CertVerifier interface {
+       VerifyPeerCertificate(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error
+}
+
 // +k8s:deepcopy-gen=true

 // ServersTransport options to configure communication between Traefik and the servers.
@@ -214,7 +218,8 @@ type ServersTransport struct {
        // order to implement the Consul Connect protocol between Traefik and the Consul
        // services. Therefore it is not serializable to prevent it from being configurable
        // by other parties.
-       VerifyPeerCertificate func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error `json:"-" toml:"-" yaml:"-" hash:"-" label:"-" export:"false"`
+       // VerifyPeerCertificate func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error `json:"-" toml:"-" yaml:"-" hash:"-" label:"-" export:"false"`
+       CertVerifier `json:"-" toml:"-" yaml:"-" hash:"-" label:"-" export:"false"`
 }

 // +k8s:deepcopy-gen=true
diff --git a/pkg/provider/consulcatalog/config.go b/pkg/provider/consulcatalog/config.go
index aeed583e9..c84bf7918 100644
--- a/pkg/provider/consulcatalog/config.go
+++ b/pkg/provider/consulcatalog/config.go
@@ -149,6 +149,24 @@ func (c *connectCert) equals(other *connectCert) bool {
        return c.leaf == other.leaf
 }

+func (dat itemData) VerifyPeerCertificate(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+       // We should use RootCAs here, but consul expect that in ClientCAs (don't ask)
+       // https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L279-L282
+       // called via https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L258
+       cfg.ClientCAs = cfg.RootCAs
+       cert, err := verifyChain(cfg, rawCerts, true)
+       if err != nil {
+               return err
+       }
+       certs := []*x509.Certificate{0: cert}
+       uri := &connect.SpiffeIDService{
+               Namespace:  dat.Namespace,
+               Datacenter: dat.Datacenter,
+               Service:    dat.Name,
+       }
+       return verifyServerCertMatchesURI(certs, uri)
+}
+
 func (c *connectCert) serverTransport(item itemData) *dynamic.ServersTransport {
        return &dynamic.ServersTransport{
                // InsecureSkipVerify is needed because Go wants to verify a hostname otherwise
@@ -157,23 +175,7 @@ func (c *connectCert) serverTransport(item itemData) *dynamic.ServersTransport {
                Certificates: tls.Certificates{
                        c.getLeaf(),
                },
-               VerifyPeerCertificate: func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-                       // We should use RootCAs here, but consul expect that in ClientCAs (don't ask)
-                       // https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L279-L282
-                       // called via https://github.com/hashicorp/consul/blob/cd428060f6547afddd9e0060c07b2a2c862da801/connect/tls.go#L258
-                       cfg.ClientCAs = cfg.RootCAs
-                       cert, err := verifyChain(cfg, rawCerts, true)
-                       if err != nil {
-                               return err
-                       }
-                       certs := []*x509.Certificate{0: cert}
-                       uri := &connect.SpiffeIDService{
-                               Namespace:  item.Namespace,
-                               Datacenter: item.Datacenter,
-                               Service:    item.Name,
-                       }
-                       return verifyServerCertMatchesURI(certs, uri)
-               },
+               CertVerifier: item,
        }
 }

diff --git a/pkg/server/service/roundtripper.go b/pkg/server/service/roundtripper.go
index fe0528250..fe503c158 100644
--- a/pkg/server/service/roundtripper.go
+++ b/pkg/server/service/roundtripper.go
@@ -146,9 +146,9 @@ func createRoundTripper(cfg *dynamic.ServersTransport) (http.RoundTripper, error
                        RootCAs:            createRootCACertPool(cfg.RootCAs),
                        Certificates:       cfg.Certificates.GetCertificates(),
                }
-               if cfg.VerifyPeerCertificate != nil {
+               if cfg.CertVerifier != nil {
                        transport.TLSClientConfig.VerifyPeerCertificate = func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
-                               return cfg.VerifyPeerCertificate(transport.TLSClientConfig, rawCerts, verifiedChains)
+                               return cfg.CertVerifier.VerifyPeerCertificate(transport.TLSClientConfig, rawCerts, verifiedChains)
                        }
                }
        }
```

Reply:
Seems to work, pushed a new commit.

Reply:
excellent, thanks.